addons/mail/models/mail_followers.py

```diff
@@ -124,6 +124,28 @@ def _get_recipient_data(self, records, message_type, subtype_id, pids=None, cids
         self.env['res.groups'].flush(['users'])
         self.env['mail.channel'].flush(['email_send', 'channel_type'])
         if records and subtype_id:
+            partner_select = """
+    SELECT partner.id as pid, NULL::int AS cid,
+            partner.active as active, partner.partner_share as pshare, NULL as ctype,
+            users.notification_type AS notif, array_agg(groups.id) AS groups
+        FROM res_partner partner
+        LEFT JOIN res_users users ON users.partner_id = partner.id AND users.active
+        LEFT JOIN res_groups_users_rel groups_rel ON groups_rel.uid = users.id
+        LEFT JOIN res_groups groups ON groups.id = groups_rel.gid
+        {join}
+        WHERE {where}
+        GROUP BY partner.id, users.notification_type
+            """
+            query_pids = ''
+            if pids:
+                query_pids = 'UNION ALL\n' + partner_select.format(join='', where='partner.id IN %s')
```

Comment:
This kind of multiple if - else format + string combination is extremely complicated to understand ... isn't it any issue with "join" that is not defined outside of if pids loop ? Or query_partner_followers called herebelow but defined here only ?

Well I have to say that this code is very complicated to read: params being extended, multiple format + %s, ... can't really say this code is bullet proof.

Reply:
I have the same feeling. I just didn't have inspiration to improve it. My original idea was to avoid  code duplicates. I'll check what could be improved